t2824: diagnose pre-push hook 60s timeout that forces --skip-hooks bypass

## Session Origin

Surfaced while triaging #20732 dispatch friction. The successful recovery worker (session that produced PR #20817) hit:

```
[ERROR] Push timed out after 60s — likely a pre-push hook stalling.
```

It worked around the timeout with `git push --no-verify` followed by `commit-and-pr --skip-hooks`. The PR landed without privacy/complexity/scope/dup-todo guard verification on the final diff. Worker log: `~/.aidevops/logs/worker-20732.log` (push-timeout block at lines 108-116, --no-verify push at line 112).

## What

Two-phase task:

1. **Phase 1 — diagnose**: instrument the four pre-push guards (`privacy-guard-pre-push.sh`, `complexity-guard-pre-push.sh`, `scope-guard-pre-push.sh`, `dup-todo-guard-pre-push.sh`) with `time` measurements and run them against the diff that triggered the original timeout. Identify which guard exceeds 60s.

2. **Phase 2 — fix**: based on Phase 1 findings, file a child issue with the targeted fix (likely one of: optimise the slow guard, raise the timeout, add a fast-path for known-safe diffs, or split the slow guard's work).

This issue covers Phase 1 only. Phase 2 is filed as a child once the diagnosis lands.

## Why

`commit-and-pr --skip-hooks` defeats the safety net the four guards exist to provide:

- **privacy-guard**: blocks private repo slugs leaking into public commits
- **complexity-guard**: blocks new function-complexity / nesting-depth / file-size violations
- **scope-guard**: blocks out-of-scope file changes per brief Files Scope
- **dup-todo-guard**: blocks duplicate task-ID checkboxes in TODO.md

When a worker bypasses with `--no-verify`, none of these run. The current escape hatch is necessary (workers must be able to push when blocked) but it's also routine — every worker that hits the timeout strips the guards from that PR. Untracked, this drifts toward "guards run nowhere".

The 60s timeout in `full-loop-helper.sh` is reasonable for a fast-path. The fix is either to make the slow guard fast, or to give it a longer budget without raising the global timeout for all four.

## How

### Phase 1 — diagnostic-first (THIS ISSUE)

Files to modify: NONE. This is pure investigation.

Steps:

1. Reproduce the timeout: check out the PR #20817 diff in a worktree, run each guard standalone:
   ```bash
   cd ~/Git/aidevops-feature-auto-20260425-004525-gh20732
   for g in privacy-guard-pre-push.sh complexity-guard-pre-push.sh scope-guard-pre-push.sh dup-todo-guard-pre-push.sh; do
       echo "=== $g ==="
       time ~/.aidevops/agents/scripts/$g
       echo
   done
   ```
2. Record the timing in this issue's body as a comment.
3. If the slow guard is reproducible, profile it (`bash -x`, identify hot loops or external calls).
4. If the slow guard is NOT reproducible (transient), add diagnostic logging to each guard so the next timeout instance leaves a breadcrumb.
5. File Phase 2 child issue with the targeted fix and link back to this issue with `Resolves` once Phase 2 lands.

### Reference patterns

- Pre-push hook architecture: `reference/pre-push-guards.md`
- Existing guard scripts: `~/.aidevops/agents/scripts/*-guard-pre-push.sh`
- Timeout location in full-loop-helper: search `Push timed out after 60s` (~line 730)

### Verification

```bash
# Phase 1 success criterion: this issue carries a comment with measured timings for all four guards on a reproducer diff.
# Phase 2 will be a separate PR with code changes.
```

### Files Scope

(Phase 1 only — investigation)

- No files modified in this phase.

## Acceptance (Phase 1 only)

1. Issue carries a comment with `time` measurements for all four pre-push guards on a reproducer diff (the #20732 diff or any large multi-file diff where the timeout has been observed).
2. The slow guard is identified by name, with its measured wall time and the dominant cost (CPU loop / external command / network / disk IO).
3. If the slow guard is reproducible, this issue also carries a brief profiling note (which line / which subprocess dominates).
4. Phase 2 child issue filed with: target file, proposed fix, estimated diff size, regression test sketch.

## Context

- Worker log evidence (PR #20817 cherry-pick session, ~21:18 UTC 2026-04-24): `~/.aidevops/logs/worker-20732.log` lines 108-120 — `commit-and-pr` push timed out at 60s, worker fell back to `git push --no-verify`, second invocation with `--skip-hooks` succeeded.
- Tier: standard. Investigation only — should not need thinking-tier.
- Not in dispatch-path file set. Auto-dispatch is fine.
- Tag as `parent-task` so the Phase 2 child can be filed cleanly underneath.

<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.10.3 plugin for [OpenCode](https://opencode.ai) v1.14.24 with claude-opus-4-7 spent 34m and 52,945 tokens on this with the user in an interactive session.

--- a/todo/tasks/GH-20835-brief.md
+++ b/todo/tasks/GH-20835-brief.md
@@ -1,0 +1,26 @@
+# GH-20835: diagnose pre-push hook 60s timeout
+
+## Session Origin
+
+Worker dispatch from issue #20835. Phase 1 investigation only — no code changes.
+
+## What
+
+Timed all four pre-push guards against a realistic diff (1 modified `.sh` file, ahead of `origin/main`).
+
+## Findings
+
+| Guard | Wall time |
+|-------|-----------|
+| `privacy-guard-pre-push.sh` | 104ms |
+| `scope-guard-pre-push.sh` | 11ms |
+| `pre-push-dup-todo-guard.sh` | 1.1s |
+| `complexity-regression-pre-push.sh` | **1m 38s** ← slow guard |
+
+Root cause: `nesting-depth` metric in `complexity-regression-helper.sh` scans all 949 `.sh` files in the repo (via `shfmt --to-json | jq` AST walk, ~50ms/file) × 2 worktrees = 97s. Exceeds the 60s timeout in `full-loop-helper.sh`.
+
+Phase 2 filed as #20842 (change-scoped scan fix).
+
+## Files Scope
+
+- `todo/tasks/GH-20835-brief.md` (this file)

